File: 2.add-two-numbers.go
/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// Input: l1 = [2,4,3], l2 = [5,6,4]
// Output: [7,0,8]
// Explanation: 342 + 465 = 807.
// Example 2:

// Input: l1 = [0], l2 = [0]
// Output: [0]
// Example 3:

// Input: l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]
// Output: [8,9,9,9,0,0,0,1]

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {

	//如果不用placeholder,需要先建立一個head存到current,才能用current.next來存下面的node
	// var head *ListNode
	// var current *ListNode

	// for _, val := range values {
	//     newNode := &ListNode{Val: val}
	//     if head == nil { // Special case: Initialize the first node
	//         head = newNode
	//         current = head
	//     } else { // General case: Attach new node to the current node
	//         current.Next = newNode
	//         current = current.Next
	//     }
	// }
	// return head

	//dummy used as a placeholder
	//The dummy node provides a consistent starting point (dummy.Next) for the list.
	//You don't need to check whether head is nil or manage separate logic for the first node.
	// This initializes a new ListNode with default values:

	// dummy.Val = 0 (default for integers in Go).
	// dummy.Next = nil (default for pointers in Go).

	dummy := &ListNode{}
	current := dummy
	carry := 0

	for l1 != nil || l2 != nil || carry != 0 {
		sum := carry
		if l1 != nil {
			//sum = sum + l1.Val
			sum += l1.Val
			//Move to the next node in l1
			l1 = l1.Next
		}
		if l2 != nil {
			//sum = sum + l2.Val
			sum += l2.Val
			//Move to the next node in l2
			l2 = l2.Next
		}

		//If the sum exceeds 10, a carry is needed. Calculate the quotient after dividing the sum by 10.
		carry = sum / 10
		//The node stored the remainder after dividing the sum by 10
		current.Next = &ListNode{Val: sum % 10}
		//Updates the current pointer to the next node.
		//End of operation logic, move to the next node.
		current = current.Next
	}

	return dummy.Next
}

// @lc code=end

File: 2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python
#
# [2] Add Two Numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        dummy_head = ListNode(0) # dummy node to simplify code
        current = dummy_head
        carry = 0

        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0

            total = val1 + val2 + carry
            carry = total // 10
            current.next = ListNode(total % 10)

            current = current.next
            if l1:l1 = l1.next
            if l2:l2 = l2.next


        return dummy_head.next


        # dummy_head = ListNode(0) # dummy node to simplify code
        # current = dummy_head
        # carry = 0 # Carry for sum > 9
      
        # # l1 = [2, 4, 3] (代表 342) , l2 = [5, 6, 4] (代表 465)
        # # 從後面到前面
        # # [7, 0, 8]  (代表 807)
        # while l1 or l2 or carry:
        #     val1 = l1.val if l1 else 0
        #     val2 = l2.val if l2 else 0

        #     # Compute sum and carry
        #     total = val1 + val2 + carry
        #     # 10 // 10 = 1  (進位)
        #     carry = total // 10 # Compute new carry
        #     # 建立 ListNode(0) -> total % 10 => 10 % 10 -> 0
        #     current.next = ListNode(total % 10) # Store the last digit

        #     #Move forward
        #     current = current.next
        #     # 如果 l1 還有下一個節點，則 l1 = l1.next，繼續遍歷
        #     # 確保 l1 不是 None
        #     if l1: l1 = l1.next
        #     if l2: l2 = l2.next

        # return dummy_head.next # Return the actual result 

# @lc code=end

